### PR TITLE
script/cpatch: initial version

### PR DESCRIPTION
**Pull request by liewegas**
Based on kubejacker.

- copy + strip (tar compress is sloooow)
- avoid touching build artifacts
- by default, patch "everything"
- optionally do actual build as root
- lump rbd and cephfs into core because mgr needs to dynamically link
  them and c++ ABI is fragile.

Caveats
- Currently assumes your local build matches the container distro (el8).  I'm using centos 8 stream

This works:
```
../src/script/cpatch -t quay.io/ceph-ci/ceph:dev --root-build
sudo ../src/cephadm/cephadm --image quay.io/ceph-ci/ceph:dev bootstrap --skip-pull ...
...
../src/script/cpatch ...
sudo systemctl ceph-$fsid.target restart
```

Also, cstart.sh and ckill.sh.  cstart.sh with no args will bring up a cluster such that ``bin/ceph -s`` will work when you're done.

### CODE REVIEW

**Comment by liewegas** [comment]
I think for a fully cephadm dev environment (no vstart!), we probably also should
- use a fixed fsid for the current build dir, and rm-cluster it before restart. maybe a ``fsid`` file in the build dir?
- a 'start' script (src/cstart.sh?) that does the fsid assignment and start? it could also assign a random tag for this build dir so that we don't collide with other checkouts on the same box
- ``--output-dir . --allow-overwrite`` so that our conf and keyring don't go into ``/etc/ceph``
- a 'stop' script (src/cstop.sh?)
- the start script could spit out the shell cpatch invocation for easy cut+paste?

**Comment by liewegas** [comment]
BTW it's pretty fast (on my box at least):
- ``--py`` is 2s (i.e., developing mgr moduels)
- ``--dashboard`` 5s
- ``--core`` is 30s (all binaries except rgw)
- ``--rgw`` is 8s
- everything (no rgs) is 37s

File: src/cstart.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    sudo ../src/script/cpatch -t $image_base:$shortid
+fi
+
+sudo ../src/cephadm/cephadm rm-cluster --force --fsid $fsid
+sudo ../src/cephadm/cephadm --image ${image_base}:${shortid} bootstrap \
+     --skip-pull \
+     --fsid $fsid \
+     --mon-addrv "[v2:$ip:$port]" \
+     --output-dir . \
+     --allow-overwrite
```
Comment (sebastian-philipp) [inline src/cstart.sh:53]: which is a downside of cstart: we need root access.

**Comment by sebastian-philipp** [comment]
I'll have a hard time running a cstart cluster, as we don't have any base containers  images for ubuntu.

File: src/ckill.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -e
+
+# fsid
+if [ -e fsid ] ; then
+    fsid=`cat fsid`
+else
+    fsid=`uuidgen`
```
Comment (dmick) [inline src/ckill.sh:7]: you can't just generate a new fsid to kill an existing cluster, right?

Comment (liewegas) [reply src/ckill.sh:7]: if you say ``rm fsid`` you'll just start up a new cephadm cluster.  the old one needs to be removed by running ``ckill.sh`` first or running ``cephadm rm-cluster --fsid ... --force``.

Comment (dmick) [reply src/ckill.sh:7]: I understand; I'm just saying, here in ckill, generating a new fsid makes no sense.  You want to find the fsid of the cluster to be killed, surely?

Comment (liewegas) [reply src/ckill.sh]: oh i see, yeah.  fixed! :man_facepalming:

**Comment by jdurgin** [comment]
seems like it'll be challenging to keep this updated with everything that gets installed - maybe we could generate cpatch with cmake, so we could rely on the install targets as the single source of truth? or using a more bare base image (with dependencies but not ceph itself installed?) so we can detect missing things more easily.

Trying to run it locally I got some errors - fixed here: https://github.com/jdurgin/ceph/commits/cpatch

I'm sure there are more missing files though.

At a higher level it behooves us to make the dev experience as close as possible to the end user experience -- so instead of ```bin/ceph```, you run ```cephadm shell ceph```, etc.

If there are things that are difficult to do, we should improve cephadm so they are easier. For example, ```cephadm shell``` could have a volume passthrough (or maybe arbitrary podman/docker option passthrough), so if you did something like ```cephadm shell --volume ~/ceph /ceph```, you'd have access to all the usual workunits and other test scripts.

We'll need an image that has the debug info and test programs just like we discussed for use in teuthology anyway - updating that from the local build tree would make a pretty complete dev environment.

Rather than requiring the host to match the container image, I think we should start building in a container instead. The initial download will be larger, but it will ease maintenance and CI resources to concentrate on a single target. I just built in a centos 8 container to get this running on ubuntu, for example, using these scripts: https://gist.github.com/jdurgin/3215df1295d7bfee0e91b203eae71dce

We can break some of the steps down so that we can save time, or run fewer to save space, e.g. creating a series of container images in ceph-ci for development. At the base would be running install-deps.sh, which takes a significant amount of time from a barebones container image. We could go so far as to do a full build - downloading a few extra gigs would be much faster than building locally in many cases. @dmick I think you've looked into this area before, what do you think?

**Comment by badone** [comment]
May also want take a look at the container specific scripts in /src/test and maybe https://github.com/ceph/ceph/pull/34327 for automating a containerised build? I just skimmed this but will take a better look on Monday.

**Comment by liewegas** [comment]
> seems like it'll be challenging to keep this updated with everything that gets installed - maybe we could generate cpatch with cmake, so we could rely on the install targets as the single source of truth? or using a more bare base image (with dependencies but not ceph itself installed?) so we can detect missing things more easily.

Probably doing something like 'make install' with DESTDIR would be the most complete, but we'd need to adjust the default build paths to match (/usr not /usr/local etc).  I'm not sure if that's possible with a subset of all the files, though.  Part of the goal here was to make this lightweight and fast.  Lots of binaries aren't included (eg osdmaptool crushtool etc).  Maybe it's not worth it though.

> Trying to run it locally I got some errors - fixed here: https://github.com/jdurgin/ceph/commits/cpatch

Cherry-picked!

> I'm sure there are more missing files though.
> 
> At a higher level it behooves us to make the dev experience as close as possible to the end user experience -- so instead of `bin/ceph`, you run `cephadm shell ceph`, etc.

:+1:
 
> If there are things that are difficult to do, we should improve cephadm so they are easier. For example, `cephadm shell` could have a volume passthrough (or maybe arbitrary podman/docker option passthrough), so if you did something like `cephadm shell --volume ~/ceph /ceph`, you'd have access to all the usual workunits and other test scripts.

:+1:

> We'll need an image that has the debug info and test programs just like we discussed for use in teuthology anyway - updating that from the local build tree would make a pretty complete dev environment.

I think @leseb is making this change to ceph-container (ceph/ceph-debug to include *-debuginfo and ceph-test[-debuginfo])
 
> Rather than requiring the host to match the container image, I think we should start building in a container instead. The initial download will be larger, but it will ease maintenance and CI resources to concentrate on a single target. I just built in a centos 8 container to get this running on ubuntu, for example, using these scripts: https://gist.github.com/jdurgin/3215df1295d7bfee0e91b203eae71dce

Yep.  I think someone just needs to add that script to he repo, and document/script how to launch it for the actual build with all the needed -v args etc.  It's a little annoying that the build results are lost if your terminal closes and the container stops.. so maybe it should bind build/ to a host dir? Or we recommend running the build inside tmux/screen or something.

> We can break some of the steps down so that we can save time, or run fewer to save space, e.g. creating a series of container images in ceph-ci for development. At the base would be running install-deps.sh, which takes a significant amount of time from a barebones container image. We could go so far as to do a full build - downloading a few extra gigs would be much faster than building locally in many cases. @dmick I think you've looked into this area before, what do you think?

We could publish the result of that Dockerfile to quay.io/ceph-ci/ceph:dev-base or something...

**Comment by jdurgin** [comment]
> > Rather than requiring the host to match the container image, I think we should start building in a container instead. The initial download will be larger, but it will ease maintenance and CI resources to concentrate on a single target. I just built in a centos 8 container to get this running on ubuntu, for example, using these scripts: https://gist.github.com/jdurgin/3215df1295d7bfee0e91b203eae71dce
> 
> Yep. I think someone just needs to add that script to he repo, and document/script how to launch it for the actual build with all the needed -v args etc. It's a little annoying that the build results are lost if your terminal closes and the container stops.. so maybe it should bind build/ to a host dir? Or we recommend running the build inside tmux/screen or something.

Yes, I was binding the entire repo including the build dir into the container so it was persistent:

https://gist.github.com/jdurgin/3215df1295d7bfee0e91b203eae71dce#file-run-docker-sh-L15

@badone points out test/docker-test.sh which may be a good building block here

> > We can break some of the steps down so that we can save time, or run fewer to save space, e.g. creating a series of container images in ceph-ci for development. At the base would be running install-deps.sh, which takes a significant amount of time from a barebones container image. We could go so far as to do a full build - downloading a few extra gigs would be much faster than building locally in many cases. @dmick I think you've looked into this area before, what do you think?
> 
> We could publish the result of that Dockerfile to quay.io/ceph-ci/ceph:dev-base or something...

Yeah, trying again with the centos base image, install-deps is much faster. I'm thinking 'make install' is going to be too slow though.

Timing was roughly:

```install-deps.sh: 7m
do_cmake.sh: 1m
make -j32: 45m
make install: 10m
make -j32 install again, w/no changes: 32s
```
Running cpatch, copying everything took 2m. I experimented more with 'make install/strip/fast' but it seems this approach of copying all the files is too slow to consider adding to a vstart replacement.

One way around this is to avoid needing to copy anything - I prototyped this by generating a list of installed files from ceph packages, and replacing them in the container image with symlinks to a volume I where I would mount the source and build tree. This has the nice property of needing no extra time (the container image does not need to change). However, it does require more work to get cephadm to include this volume in every container it launches. Since cephadm runs as root, this requires making them privileged containers as well, to be able access my normal-user checkout. Perhaps there's a more fine-grained way to allow access. This does make a dev setup a bit more distinct from normal use of cephadm though.

In any case, the approach of figuring out which files to symlink via package inspection is very fast, and could be applied to the cpatch approach, copying only pieces you're developing, as well.

Scripts are here, cephadm requires changes to work with this approach still:
https://gist.github.com/jdurgin/3215df1295d7bfee0e91b203eae71dce#file-run-prep-container-sh
https://gist.github.com/jdurgin/3215df1295d7bfee0e91b203eae71dce#file-prepare-dev-container-sh

What do folks think, is it worth going down the path of an immutable container dev image at the cost of more cephadm changes to accomadate a dev environment?

Or should we stick to copying like cpatch, adding a bit more time to spin up a cluster?

**Comment by jdurgin** [comment]
Discussed with @sebastian-philipp this morning, it seems like the changes needed for cephadm will not be too invasive if we continue running as root.

We talked about using environment variables set by cstart.sh to tell cephadm to operate in dev mode and mount the ceph repo as a volume.

**Comment by liewegas** [comment]
> Discussed with @sebastian-philipp this morning, it seems like the changes needed for cephadm will not be too invasive if we continue running as root.
> 
> We talked about using environment variables set by cstart.sh to tell cephadm to operate in dev mode and mount the ceph repo as a volume.

My only concern with this approach is it prevents you from testing/deploying across multiple hosts. I think one way or another we need a way to produce a "real" container image, even if there is a speedier workflow for developers.  We just need to balance that speed vs maintaining multiple ways to work

**Comment by jdurgin** [comment]
> > Discussed with @sebastian-philipp this morning, it seems like the changes needed for cephadm will not be too invasive if we continue running as root.
> > We talked about using environment variables set by cstart.sh to tell cephadm to operate in dev mode and mount the ceph repo as a volume.
> 
> My only concern with this approach is it prevents you from testing/deploying across multiple hosts. I think one way or another we need a way to produce a "real" container image, even if there is a speedier workflow for developers. We just need to balance that speed vs maintaining multiple ways to work

How often is that required from a dev environment? You could use a shared fs for your repository to access across nodes... or you could wait a couple hours using ceph-ci to make a full container.

**Comment by liewegas** [comment]
> > > Discussed with @sebastian-philipp this morning, it seems like the changes needed for cephadm will not be too invasive if we continue running as root.
> > > We talked about using environment variables set by cstart.sh to tell cephadm to operate in dev mode and mount the ceph repo as a volume.
> > 
> > 
> > My only concern with this approach is it prevents you from testing/deploying across multiple hosts. I think one way or another we need a way to produce a "real" container image, even if there is a speedier workflow for developers. We just need to balance that speed vs maintaining multiple ways to work
> 
> How often is that required from a dev environment? You could use a shared fs for your repository to access across nodes... or you could wait a couple hours using ceph-ci to make a full container.

I think it will come up pretty often, especially with newer work (haproxy support, using this for rgw/cephfs/rbd multisite, etc.).  ~2m is a big improvement from ~1h (and way more resource efficient)

**Comment by sebastian-philipp** [comment]
> > > > Discussed with @sebastian-philipp this morning, it seems like the changes needed for cephadm will not be too invasive if we continue running as root.
> > > > We talked about using environment variables set by cstart.sh to tell cephadm to operate in dev mode and mount the ceph repo as a volume.
> > > 
> > > 
> > > My only concern with this approach is it prevents you from testing/deploying across multiple hosts. I think one way or another we need a way to produce a "real" container image, even if there is a speedier workflow for developers. We just need to balance that speed vs maintaining multiple ways to work
> > 
> > 
> > How often is that required from a dev environment? You could use a shared fs for your repository to access across nodes... or you could wait a couple hours using ceph-ci to make a full container.
> 
> I think it will come up pretty often, especially with newer work (haproxy support, using this for rgw/cephfs/rbd multisite, etc.). ~2m is a big improvement from ~1h (and way more resource efficient)

At least in my experience developing mgr/rook with Johns kubejacker, it was a pain to work with, as all changes took an awful long time to be present. At least I enjoy the direct feedback I get from calling `mgr fail` without the need to wait 30s.

**Comment by liewegas** [comment]
Should we just merge this so that it is an option in the tree?  I don't think we're going to settle on a be-all end-all solution in any single PR

File: src/cstart.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    while [ true ]
+    do
+        port="$(echo $(( RANDOM % 1000 + 40000 )))"
+        ss -a -n | grep LISTEN | grep "${ip}:${port} " 1>/dev/null 2>&1 || break
+    done
+fi
+echo "port $port"
+
+# make sure we have an image
+if ! podman image inspect $image_base:$shortid 2>/dev/null; then
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/cstart.sh:42]: fixed!

**Pull request merged**